projects/optic/src/commands/api/add.ts

```diff
@@ -0,0 +1,215 @@
+    .option(
+      '--history-depth <history-depth>',
+      'Set rulesets to run on API diffs. You can always add this later by setting the `[x-optic-ruleset]` key on your OpenAPI spec',
+      '1'
+    )
+    .option(
+      '--ruleset <ruleset>',
+      'Set rulesets to run on API diffs. You can always add this later by setting the `[x-optic-ruleset]` key on your OpenAPI spec'
+    )
+    .option('--web', 'open to the added API in Optic Cloud', false)
+    .action(wrapActionHandlerWithSentry(getApiAddAction(config)));
+};
+
+type ApiAddActionOptions = {
+  historyDepth: string;
+  ruleset?: boolean;
+  web?: boolean;
+};
+
+const getApiAddAction =
+  (config: OpticCliConfig) =>
+  async (path_to_spec: string | undefined, options: ApiAddActionOptions) => {
+    if (isNaN(Number(options.historyDepth))) {
+      logger.error(
+        '--history-depth is not a number. history-depth must be a number'
+      );
+      process.exitCode = 1;
+      return;
+    } else if (!path_to_spec && options.historyDepth !== '1') {
+      logger.error(
+        'Invalid argument combination: Cannot set a history-depth !== 1 when no spec path is provided'
+      );
+      process.exitCode = 1;
+      return;
+    } else if (!config.isAuthenticated) {
+      logger.error('Must be logged in to add APIs. Log in with `optic login`');
+      process.exitCode = 1;
+      return;
+    } else if (!config.vcs || config.vcs?.type !== VCS.Git) {
+      logger.error('Must have git in path and be in a git repo to add apis');
+      process.exitCode = 1;
+      return;
+    }
+
+    if ('todo fetch orgs related to token') {
+      //  - if more than one org, ask for which org to add to
+    }
+
+    if (!options.ruleset) {
+      console.log('TODO set info for rulesets');
+      //  - prompt for a ruleset / give suggestions
+    }
+
+    const candidates = path_to_spec
+      ? await getShasCandidatesForPath(path_to_spec, options.historyDepth)
+      : await getPathCandidatesForSha(config.vcs.sha);
+
+    for await (const [path, shas] of candidates) {
+      const [firstSha] = shas;
+
+      let parseResult: ParseResult;
+      try {
+        parseResult = await getFileFromFsOrGit(`${firstSha}:${path}`, config);
+      } catch (e) {
+        if (path === path_to_spec) {
+          logger.info(
+            `File ${path_to_spec} is not a valid OpenAPI file. Optic currently supports OpenAPI 3 and 3.1`
+          );
+          logger.info(e);
+        } else {
+          logger.debug(`Disregarding candidate ${path}`);
+          logger.debug(e);
+        }
+        continue;
+      }
+      if (parseResult.isEmptySpec) {
+        logger.info(`File ${path} does not exist in sha ${firstSha}`);
+        continue;
+      }
+
+      const opticUrl = parseResult.jsonLike[OPTIC_URL_KEY];
+      const apiId = opticUrl
+        ? 'todo get optic id from url'
+        : 'todo make API call';
+
+      logger.info(`Found OpenAPI Spec at ${path}`);
+      for await (const sha of shas) {
```

Comment:
Here we traverse the history until we hit:
- an invalid API
- an empty spec

I'm not convinced this is correct, and maybe instead we should just ignore the empty / invalid specs and continue, instead of stopping. Some valid cases:
- Someone uploads an invalid API, and then fixes it in a follow up PR
- Someone deletes and API and then reverts the changes

Reply:
I just dug through detecting renames and I don't think it's worth the effort. So yeah, I'd say just stop when you get to the first copy of the spec, and hopefully teams don't rename them too terribly often. We could write some kind of plumbing spec upload code for teams that want to walk renames.